File: examples/cpp/ToF/tof_vd55h1_controls.cpp
#include <array>
#include <memory>
#include <opencv2/opencv.hpp>

#include "depthai/depthai.hpp"

namespace {

constexpr float FPS = 30.0f;
constexpr const char* WINDOW = "VD55H1 controls";

std::shared_ptr<dai::ToFConfig> configFromTrackbars(const std::array<int, 10>& value) {
    auto config = std::make_shared<dai::ToFConfig>();
    auto& vd55h1 = config->vd55h1;
    vd55h1.phaseUnwrapErrorThreshold = value[0];
    vd55h1.enableBilateralFilter = value[1] != 0;
    vd55h1.bilateralStdFactor = value[2] / 100.0f;
    vd55h1.bilateralKernelSize = value[3];
    vd55h1.enableTemporalNoiseReduction = value[4] != 0;
    vd55h1.temporalNoiseReductionMaxGain = value[5];
    vd55h1.temporalNoiseReductionStdFactor = value[6] / 100.0f;
    vd55h1.enableFlyingPixelFilter = value[7] != 0;
    vd55h1.flyingPixelDepthThreshold = value[8];
    vd55h1.flyingPixelMinDepthOccurrence = value[9] / 100.0f;
    return config;
}

}  // namespace

int main() {
    dai::Pipeline pipeline;
    auto tof = pipeline.create<dai::node::ToF>()->build(dai::CameraBoardSocket::AUTO, dai::ToFConfig::Profile::MID_RANGE, FPS);
    auto depthQueue = tof->depth.createOutputQueue(1, false);
    auto configQueue = tof->tofBaseInputConfig.createInputQueue();

    cv::namedWindow(WINDOW);
    std::array<int, 10> value = {192, 1, 205, 5, 1, 27, 82, 1, 101, 1356};
    cv::createTrackbar("unwrap threshold", WINDOW, &value[0], 500);
    cv::createTrackbar("bilateral", WINDOW, &value[1], 1);
    cv::createTrackbar("bilateral std x100", WINDOW, &value[2], 1000);
    cv::createTrackbar("bilateral kernel", WINDOW, &value[3], 15);
    cv::createTrackbar("temporal NR", WINDOW, &value[4], 1);
    cv::createTrackbar("TNR max gain", WINDOW, &value[5], 100);
    cv::createTrackbar("TNR std x100", WINDOW, &value[6], 500);
    cv::createTrackbar("flying pixel", WINDOW, &value[7], 1);
    cv::createTrackbar("FP depth threshold", WINDOW, &value[8], 1000);
    cv::createTrackbar("FP min occurrence x100", WINDOW, &value[9], 5000);

    pipeline.start();
    std::array<int, 10> previous{};
    previous.fill(-1);
    while(pipeline.isRunning()) {
        if(value != previous) {
            configQueue->send(configFromTrackbars(value));
            previous = value;
        }

        if(const auto frame = depthQueue->tryGet<dai::ImgFrame>()) {
            cv::imshow("ToF depth", dai::utility::colorizeDepthFrame(*frame, 100.0f, 7000.0f, cv::COLORMAP_JET, true).getCvFrame());
        }
        if(cv::waitKey(1) == 'q') break;
    }
    return 0;
}

File: examples/python/ToF/tof_vd55h1_controls.py
#!/usr/bin/env python3
"""Tune every RVC4 VD55H1 ToF IPP control with OpenCV sliders.

Trackbars ending in "x100" use a scale factor of 100. Toggle trackbars use
0 for off and 1 for on. Press q to quit.
"""

import cv2
import depthai as dai


FPS = 30.0
WINDOW = "VD55H1 controls"


def config_from_trackbars() -> dai.ToFConfig:
    config = dai.ToFConfig()
    vd55h1 = config.vd55h1
    vd55h1.phaseUnwrapErrorThreshold = cv2.getTrackbarPos("unwrap threshold", WINDOW)
    vd55h1.enableBilateralFilter = bool(cv2.getTrackbarPos("bilateral", WINDOW))
    vd55h1.bilateralStdFactor = cv2.getTrackbarPos("bilateral std x100", WINDOW) / 100.0
    vd55h1.bilateralKernelSize = cv2.getTrackbarPos("bilateral kernel", WINDOW)
    vd55h1.enableTemporalNoiseReduction = bool(cv2.getTrackbarPos("temporal NR", WINDOW))
    vd55h1.temporalNoiseReductionMaxGain = cv2.getTrackbarPos("TNR max gain", WINDOW)
    vd55h1.temporalNoiseReductionStdFactor = cv2.getTrackbarPos("TNR std x100", WINDOW) / 100.0
    vd55h1.enableFlyingPixelFilter = bool(cv2.getTrackbarPos("flying pixel", WINDOW))
    vd55h1.flyingPixelDepthThreshold = cv2.getTrackbarPos("FP depth threshold", WINDOW)
    vd55h1.flyingPixelMinDepthOccurrence = cv2.getTrackbarPos("FP min occurrence x100", WINDOW) / 100.0
    vd55h1.enableRadialToPerpendicularCorrection = bool(cv2.getTrackbarPos("IPP radial-to-perp", WINDOW))
    return config


def main() -> None:
    with dai.Pipeline() as pipeline:
        tof = pipeline.create(dai.node.ToF).build(
            boardSocket=dai.CameraBoardSocket.AUTO,
            profile=dai.ToFConfig.Profile.MID_RANGE,
            fps=FPS,
        )
        depth_queue = tof.depth.createOutputQueue(maxSize=1, blocking=False)
        config_queue = tof.tofBaseInputConfig.createInputQueue()

        cv2.namedWindow(WINDOW)
        controls = (
            ("unwrap threshold", 500, 192),
            ("bilateral", 1, 1),
            ("bilateral std x100", 1000, 205),
            ("bilateral kernel", 15, 5),
            ("temporal NR", 1, 1),
            ("TNR max gain", 100, 27),
            ("TNR std x100", 500, 82),
            ("flying pixel", 1, 1),
            ("FP depth threshold", 1000, 101),
            ("FP min occurrence x100", 5000, 1356),
            ("IPP radial-to-perp", 1, 0),
        )
        for name, maximum, initial in controls:
            cv2.createTrackbar(name, WINDOW, initial, maximum, lambda _: None)

        pipeline.start()
        previous = None
        while pipeline.isRunning():
            values = tuple(cv2.getTrackbarPos(name, WINDOW) for name, _, _ in controls)
            if values != previous:
                config_queue.send(config_from_trackbars())
                previous = values

            frame = depth_queue.tryGet()
            if frame is not None:
                cv2.imshow("ToF depth", dai.utility.colorizeDepthFrame(frame, useLog=True).getCvFrame())

            if cv2.waitKey(1) == ord("q"):
                break

    cv2.destroyAllWindows()


if __name__ == "__main__":
    main()
